trie/triedb/pathdb/disklayer.go

```diff
@@ -0,0 +1,259 @@
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package pathdb
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/trie/trienode"
+	"github.com/ethereum/go-ethereum/trie/triestate"
+)
+
+// diskLayer is a low level persistent layer built on top of a key-value store.
+type diskLayer struct {
+	rootHash common.Hash  // Immutable, root hash of the base layer
+	id       uint64       // Immutable, corresponding state id
+	db       *Database    // Path-based trie database
+	buffer   *nodebuffer  // Node buffer to aggregate writes.
+	stale    bool         // Signals that the layer became stale (state progressed)
+	lock     sync.RWMutex // Lock used to protect stale flag
+}
+
+// newDiskLayer creates a new disk layer based on the passing arguments.
+func newDiskLayer(root common.Hash, id uint64, db *Database, buffer *nodebuffer) *diskLayer {
+	return &diskLayer{
+		rootHash: root,
+		id:       id,
+		db:       db,
+		buffer:   buffer,
+	}
+}
+
+// root returns root hash of corresponding state.
+func (dl *diskLayer) root() common.Hash {
+	return dl.rootHash
+}
+
+// parent always returns nil as there's no layer below the disk.
+func (dl *diskLayer) parent() layer {
+	return nil
+}
+
+// stateID returns the state id of disk layer.
+func (dl *diskLayer) stateID() uint64 {
+	return dl.id
+}
+
+// isStale return whether this layer has become stale (was flattened across) or if
+// it's still live.
+func (dl *diskLayer) isStale() bool {
+	dl.lock.RLock()
+	defer dl.lock.RUnlock()
+
+	return dl.stale
+}
+
+// markStale sets the stale flag as true.
+func (dl *diskLayer) markStale() {
+	dl.lock.Lock()
+	defer dl.lock.Unlock()
+
+	if dl.stale {
+		panic("triedb disk layer is stale") // we've committed into the same base from two children, boom
+	}
+	dl.stale = true
+}
+
+// Node retrieves the trie node with the provided node info. No error will be
+// returned if the node is not found.
+func (dl *diskLayer) Node(owner common.Hash, path []byte, hash common.Hash) ([]byte, error) {
+	dl.lock.RLock()
+	defer dl.lock.RUnlock()
+
+	if dl.stale {
+		return nil, errSnapshotStale
+	}
+	// Try to retrieve the trie node from the not-yet-written
+	// node buffer first. Note the buffer is lock free since
+	// it's impossible to mutate the buffer before tagging the
+	// layer as stale.
+	n, err := dl.buffer.node(owner, path, hash)
+	if err != nil {
+		return nil, err
+	}
+	if n != nil {
+		dirtyHitMeter.Mark(1)
+		dirtyReadMeter.Mark(int64(len(n.Blob)))
+		return n.Blob, nil
+	}
+	dirtyMissMeter.Mark(1)
+
+	// Try to retrieve the trie node from the clean memory cache
+	if dl.db.cleans != nil {
+		if blob := dl.db.cleans.Get(nil, hash.Bytes()); len(blob) > 0 {
```

Comment:
Honestly, I think i was wrong here by caching nodes via hash.

(1) clean cache is supposed to a hot cache of persistent state,
it must only contain the nodes which are present in disk

(2) cache nodes by hash means the stale nodes won't be overwritten
and stay there until cache eviction. Theoretically it's possible 
we can retrieve a stale node which is not a part of persistent state
but accidentally hit in the cache.

Originally I choose hash is because it's easier to handle than path.
but yeah, i will try to fix it.

Reply:
The strange issue with caching by hash is that you could maybe end up in a weird deduplicated data scenario where something is in the cache but not on the path you're currently touching (e.g. 2 contracts having the same state). By path would ensure that deduplication doens't happen in cache either, but then we need to be really careful with evicting old cache entries on both commits and revertals.

Reply:
Exactly, we need to be really careful for manipulating the clean cache, otherwise can end up in a bad state situation.

For the operations supported by pathdb:
- Commit:
   the content in the nodeBuffer(disk layer) is flushed into disk, we need to insert items into clean cache and delete items from cache if the node is marked as deleted
 
- Revert:
   the persisted state is reverted by a reverse diff, we need to insert items into clean cache and delete items from cache if the node is marked as deleted

- Reset:
   a special operation which clear all states in pathdb and revert to null, In this case, the entire clean cache needs to be reset as well